### PR TITLE
Added additional stats to Plugin interface

### PR DESCRIPTION
**Pull request by AT41**

### CODE REVIEW

File: distribution/openrct2.d.ts
Comment (Broxzier) [inline distribution/openrct2.d.ts:2083]: Wouldn't it be nicer to return the speed in miles per hour here? I guess it wouldn't be as accurate. Perhaps a custom `Speed` type would be even better.

**Review by Broxzier — CHANGES_REQUESTED** [review]
I'd rather see the getters throw when `ride` is null, instead of returning a magic number. This can be done later in another PR, just please create an issue for it if you decide to leave it for now.

Please increment the API version, then this is good to go.

File: distribution/openrct2.d.ts
Comment (AT41) [reply distribution/openrct2.d.ts:2083]: There's a lot of these unreadable values that are coming in, each with their own formulas, and some are in metric while others are imperial. From a plugin dev pov I think it would be easier if I just converted everything to metric here.

Comment (AT41) [reply distribution/openrct2.d.ts:2083]: @Sadret initially wanted them returned in their original form and letting the plugin devs format them. Thoughts?

Comment (Sadret) [reply distribution/openrct2.d.ts:2083]: In my opinion, the best solution is to have all stats scaled such that they can be used within the `formatString` method, in this case with `"{VELOCITY}"`.

**Comment by ZehMatt** [comment]
> I'd rather see the getters throw when `ride` is null, instead of returning a magic number. This can be done later in another PR, just please create an issue for it if you decide to leave it for now.
> 
> Please increment the API version, then this is good to go.

I think the error handling should be changed in this PR already, that's quite the different behavior.

**Comment by Broxzier** [comment]
> I think the error handling should be changed in this PR already, that's quite the different behavior.

It's not different from the current behaviour. The existing ride properties don't throw on null either, so it needs to be done for all the other ride properties too, then.

**Comment by ZehMatt** [comment]
> > I think the error handling should be changed in this PR already, that's quite the different behavior.
> 
> It's not different from the current behaviour. The existing ride properties don't throw on null either, so it needs to be done for all the other ride properties too, then.

Fair point.

**Comment by AT41** [comment]
API incremented, modified return values so that they can be formatted through {VELOCITY} and {LENGTH}. Will make a new issue for throwing exceptions for getters rather than returning 0 once PR is accepted.

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (Broxzier) [inline src/openrct2/scripting/bindings/ride/ScRide.cpp:572]: This is blocking the CI from creating build artifacts:
```suggestion
        return ride != nullptr ? ride->total_air_time * 3 : 0;
```

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (Basssiiie) [inline src/openrct2/scripting/bindings/ride/ScRide.cpp:548]: Aren't we loosing some accuracy here either?

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (Basssiiie) [inline src/openrct2/scripting/bindings/ride/ScRide.cpp:590]: Aren't we loosing some accuracy here as well?

File: distribution/openrct2.d.ts
Comment (Sadret) [inline distribution/openrct2.d.ts:2088]: `in miles per hour`

File: distribution/openrct2.d.ts
Comment (Sadret) [inline distribution/openrct2.d.ts:2101]: gram???

File: distribution/openrct2.d.ts
Comment (Sadret) [inline distribution/openrct2.d.ts:2106]: same

File: distribution/openrct2.d.ts
Comment (Sadret) [inline distribution/openrct2.d.ts:2116]: `The total airtime in hundredths of a second.`
No multiplication needed.

File: distribution/openrct2.d.ts
Comment (Sadret) [inline distribution/openrct2.d.ts:2096]: `in metres`

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (Sadret) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:548]: Return value is in meters, so we are loosing sub-meter accuracy. But it yields the value which is used in the ride stats tab, so I think this is fine.

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (Sadret) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:590]: Same as above.

**Review by Sadret — CHANGES_REQUESTED** [review]

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:548]: It makes it harder to use the values for other purposes though, like if a plugin want to mimic sheltered surfaces with the total ride length or use the `highest_drop_height` for calculating certain ride ratings. Also note that internally `highest_drop_height` is in height units, which is a perfectly valid measurement used in other parts of the plugin API.

File: distribution/openrct2.d.ts
Comment (AT41) [reply distribution/openrct2.d.ts:2096]: Forgot to check the typings in, oops!

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (AT41) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:548]: We're kind of running into the same issue with Max Speed and Average Speed as well, so should we not modify those either? If there are specific use cases where we can't use these values, I can make `rideLengthUnmodified_get()` functions as well.

Comment (Sadret) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:548]: On of the ideas of the plugin API design is to hide how things internally work. That also means that we should not provide raw values, especially since they might me subject of change.
If higher precision is absolutely needed, I would go for milliseconds or real values.

Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:548]: Or if possible, maybe add the extra detail as a floating point? Not sure how the `formatString` will interpret that though.

Comment (Sadret) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:548]: That's what I said, no?

Comment (Sadret) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:548]: image
`formatString` seems to have no issue with that.

Comment (AT41) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:548]: Sounds good, I'll return these values with higher precision then, maybe into the thousandths.

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (Basssiiie) [inline src/openrct2/scripting/bindings/ride/ScRide.cpp:590]: Maybe have this one as integers without any calculations, as it's the height in units which is also what the other height properties in the API return.

Comment (AT41) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:590]: Sure! I'm thinking of adding another case to formatString() to turn height units into their respective imperial / metric equivalents so plugin modders don't need to do that manually. Anything else that I'm missing?

**Review by Basssiiie — APPROVED** [review]
LGTM :)

File: src/openrct2/util/Util.cpp
Comment (Sadret) [inline src/openrct2/util/Util.cpp:56]: `unit`

File: distribution/openrct2.d.ts
Comment (Sadret) [inline distribution/openrct2.d.ts:2103]: Hundredths of G is a weird unit. Since we now have floats for the speed stats, why not also have them for the force stats?

File: distribution/openrct2.d.ts
Comment (Sadret) [inline distribution/openrct2.d.ts:2118]: Same as above.

File: distribution/openrct2.d.ts
Comment (Sadret) [inline distribution/openrct2.d.ts:2131]: Why explain it here but nowhere else? E.g. for the speed.

**Review by Sadret — CHANGES_REQUESTED** [review]

File: distribution/openrct2.d.ts
Comment (AT41) [reply distribution/openrct2.d.ts:2103]: I figured that since it's ok for excitement/nausea/intensity to get returned as hundredths then it'd be fine if Gs were returned as hundredths as well, but for my uses it'd be more convenient to return as a float anyways. Will change.

File: distribution/openrct2.d.ts
Comment (AT41) [reply distribution/openrct2.d.ts:2131]: I feel that formatString() isn't that well documented anywhere, so I added my own comments for the one I implemented. Sure, I can add it for the rest of the Ride interface, but it probably should've been there to begin with.

**Comment by StephanSpengler** [comment]
About the formatting comments:
First, I think it is cluttering the documentation. In my opinion we do not need to cross-reference every function that might be useful in some context.
Second, formatString does not "convert to kph" but to a string representation of a localised value and unit. Plugin developers should code in a way that is independent of the actual units. formatString should not be used to convert internal units to some other units that the dev prefers to work with. The documentation should be clear to not suggest that.

All in all, my proposal is to improve the formatString documentation instead (separate PR).

**Comment by Basssiiie** [comment]
Please keep open.

**Comment by SpartanFrederic104** [comment]
Again, please keep this one open.

**Comment by ZehMatt** [comment]
@AT41 are you planning to continue this PR?

**Comment by Basssiiie** [comment]
Are there still any unresolved issues? It seems @Sadret's last comment was addressed in a commit without another comment on the PR.

**Comment by AT41** [comment]
Sorry, been busy but I'll see this PR through to the end!

All the requests should be resolved in the latest version - I'll add a new issue for throwing exceptions for null rides. Comments should also be good unless @Sadret has any further suggestions.

There's a CI/CD error that's failing after I implemented the extra enum for formatting, but unless there's a specific reason for those enum values to be set to those values the tests just need to be adjusted.
`OpenRCT2\test\tests\FormattingTests.cpp(46): error: Expected equality of these values:   "[29:{BLACK}][1:Guests: ][8:{INT32}]"   actual     Which is: "[30:{BLACK}][1:Guests: ][8:{INT32}]"`

Let me know if there's anything else that's needed on my end, thanks!

**Comment by ZehMatt** [comment]
> There's a CI/CD error that's failing after I implemented the extra enum for formatting, but unless there's a specific reason for those enum values to be set to those values the tests just need to be adjusted. `OpenRCT2\test\tests\FormattingTests.cpp(46): error: Expected equality of these values: "[29:{BLACK}][1:Guests: ][8:{INT32}]" actual Which is: "[30:{BLACK}][1:Guests: ][8:{INT32}]"`
> 
> Let me know if there's anything else that's needed on my end, thanks!

It would be good if you can change the tests if they need adjustments.

**Comment by AT41** [comment]
Ready for another review @Sadret @Broxzier

File: src/openrct2/localisation/Formatting.cpp
Comment (Sadret) [inline src/openrct2/localisation/Formatting.cpp:582]: If `gConfigGeneral.ShowHeightAsUnits` is enabled, what does the ride stat window show? Does it show the drop height as metres/feet or units? If it is the latter, this should be reflected here.

File: distribution/openrct2.d.ts
Comment (Sadret) [inline distribution/openrct2.d.ts:2124]: The total airtime in seconds.

**Review by Sadret — CHANGES_REQUESTED** [review]

File: src/openrct2/localisation/Formatting.cpp
Comment (AT41) [reply src/openrct2/localisation/Formatting.cpp:582]: The ride stat window doesn't show height measurements in units, only metres/feet.

**Comment by AT41** [comment]
Time for another review :)

**Comment by tupaschoal** [comment]
@Broxzier, @Sadret and @Basssiiie would you give this another review, please?

**Review by Basssiiie — APPROVED** [review]
The newest changes look good to me, I've got no further feedback. :)

**Review by Sadret — APPROVED** [review]
LGTM

**Comment by tupaschoal** [comment]
Seems to be an error on the mingw build, can you check @AT41 ?

**Comment by AT41** [comment]
@Broxzier Everything seems to be good to go. Added a new issue as you suggested earlier: https://github.com/OpenRCT2/OpenRCT2/issues/20994

Let me know if there's anything else needed

**Review by Broxzier — APPROVED** [review]

**Comment by tupaschoal** [comment]
@AT41 can you add a changelog entry for this, please?

File: distribution/changelog.txt
Comment (Broxzier) [inline distribution/changelog.txt:6]:
```suggestion
- Feature: [#20539] [Plugin] Add additional ride stats to plugin interface.
```

**Review by tupaschoal — APPROVED** [review]
Haven't tested, but code-wise seems ok

**Comment by Gymnasiast** [comment]
Needs updating for the changelog.

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (Gymnasiast) [inline src/openrct2/scripting/bindings/ride/ScRide.cpp:530]: Looks like this is copy-pasted? There should be a function available for this sort of thing.

**Review by Gymnasiast — CHANGES_REQUESTED** [review]

**Comment by pfroud** [comment]
I think this pull request is to support new features in the [StatRequirementChecklist](https://github.com/AT41/StatRequirementChecklist) plugin.

You may want to coordinate with pull request #20913, which I think has a similar goal in mind.

I think that PR would let you read the requirements from the native openrct2 code, so you wouldn't need to hard-code them in the plugin.

**Comment by Basssiiie** [comment]
I've tested the implementation and while most of it looks great and correct.

Two issues though;
- The formatted height string of the highest drop height seems to be off by one in some rare cases, like with the Giga Coaster in Six Flags Holland and this blue coaster I built (it's not launched so it doesn't make a full lap).
- I cannot figure out what the "powered lifts" are supposed to refer to. For some rides it's 0 and for some it's 64 (which is either incorrect or simply a flag instead of a number?):

image
image


On other rides everything seems right:

image



<details>
<summary>Click here to see my plugin script in case anyone would like to test it for themselves. 🙂 </summary>

```js
registerPlugin({
	name: "Test for ride stats",
	version: "1",
	authors: ['Basssiiie'],
	targetApiVersion: 82,
	type: "local",
	licence: "MIT",
	main() {
		ui.registerMenuItem("Check ride stats", function() {
			var dropdownName = "test-ride-stats-dropdown";
			var listName = "test-ride-stats-list";
			var rides = map.rides.sort(function(a,b){return a.name.localeCompare(b.name)});
			var window = ui.openWindow({
				classification: "test-ride-stats",
				title: "Ride stats",
				width: 300,
				height: 220,
				widgets: [
					{
						type: "listview",
						name: listName,
						showColumnHeaders: true,
						isStriped: true,
						columns: [
							{ header: "Property", width: 150 },
							{ header: "Raw", width: 40 },
							{ header: "Real", width: 80 },
						],
						items: [],
						x: 5,
						y: 20,
						width: 290,
						height: 170,
					},
					{
						type: "dropdown",
						name: dropdownName,
						x: 5,
						y: 200,
						width: 290,
						height: 15,
						items: rides.map(function(r){return r.name})
					},
				],
				onUpdate()
				{
					var dropdown = window.findWidget(dropdownName);
					var ride = rides[dropdown.selectedIndex];
					if (!ride) return;

					var list = window.findWidget(listName);
					list.items = [
						[ "Max speed:", String(ride.maxSpeed), context.formatString('{VELOCITY}', ride.maxSpeed) ],
						[ "Average speed:", String(ride.averageSpeed), context.formatString('{VELOCITY}', ride.averageSpeed) ],
						[ "Ride time:", String(ride.rideTime), context.formatString('{DURATION}', ride.rideTime) ],
						[ "Ride length:", String(ride.rideLength), context.formatString('{LENGTH}', ride.rideLength) ],
						[ "Max positive vertical Gs:", String(ride.maxPositiveVerticalGs), "" ],
						[ "Max negative vertical Gs:", String(ride.maxNegativeVerticalGs), "" ],
						[ "Max lateral Gs:", String(ride.maxLateralGs), "" ],
						[ "Total air time:", String(ride.totalAirTime), "" ],
						[ "Drops:", String(ride.drops), "" ],
						[ "Powered lifts:", String(ride.poweredLifts), "" ],
						[ "Highest drop height (raw):", String(ride.highestDropHeight), context.formatString('{HEIGHT}', ride.highestDropHeight) ]
					];
				}
			})
		})
	}
});
```

</details>

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (AT41) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:530]: You're talking about the calculation for speed I presume. From what I can see, there's a couple of places that do have the same calculation like openrct2-ui/windows/InstallTrack.cpp but they're few and far in between. I think it's fine as is since the calc is small, but do you know of a util class I can refactor them into?

**Comment by AT41** [comment]
Fixed the bug with height and powered lifts. The bit logic on height caused precision to be lost, but the new calculations should fix that. Powered lift values are combined with number of drops in a byte but I didn't truncate off the trailing 0s, so that should be fixed as well

**Review by Basssiiie — APPROVED** [review]
I've tested the fixes again. It seems powered lifts is just the count of lift hills on the ride? And it seems to be capped at a maximum of 3 regardless of this PR. Maybe an idea to add a note in the comments for this?

Either way, this PR looks good to me. :)

**Review by tupaschoal — APPROVED** [review]
@Broxzier @Gymnasiast want to do another round of review?

File: distribution/changelog.txt
Comment (tupaschoal) [inline distribution/changelog.txt:6]: This needs to be rebased to be in the 0.4.8 changelog entry btw

**Comment by Gymnasiast** [comment]
@tupaschoal Yes, I’d like to do another round.

**Comment by AT41** [comment]
@Gymnasiast Is there anything blocking this PR?

File: src/openrct2/localisation/FormatCodes.cpp
Comment (Gymnasiast) [inline src/openrct2/localisation/FormatCodes.cpp:48]: Could you explain this addition? It is unusual that a plugin PR would need to add a formatting token.

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (Gymnasiast) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp:530]: The Util class where you already added a function will do. I want to avoid copypasting, even in this file alone it is already copied twice. The same applies to the other things here. (We are still suffering from bad code quality from the early days, so I don’t want to make it worse.)

File: src/openrct2/localisation/FormatCodes.cpp
Comment (AT41) [reply src/openrct2/localisation/FormatCodes.cpp:48]: There was a conversation a while ago where we were talking about what units to return for height, since the game stores the height in height units. I'm pretty sure we decided to return these as height units and let scripters change them through the formatting function, since we return height units elsewhere (see [https://github.com/OpenRCT2/OpenRCT2/pull/20539#discussion_r1259926459](discussion))

Comment (AT41) [reply src/openrct2/localisation/FormatCodes.cpp:48]: This PR has been ongoing for a REALLY long time for a relatively simple addition, is this + [this comment](https://github.com/OpenRCT2/OpenRCT2/pull/20539#discussion_r1545492165) the last steps needed before being merged?

I barely even play the game anymore 😂

Comment (Gymnasiast) [reply src/openrct2/localisation/FormatCodes.cpp:48]: Sorry for my late reply, we are considerably understaffed and reviewing third-party PRs often falls through the cracks. I’ll try and take a look after the v0.4.11 release.

**Comment by tupaschoal** [comment]
Needs updating from 30 to 31 on https://github.com/AT41/OpenRCT2/blob/develop/test%2Ftests%2FFormattingTests.cpp

File: distribution/changelog.txt
Comment (tupaschoal) [reply distribution/changelog.txt:5]:
```suggestion
- Feature: [#20539] [Plugin] Add API for getting a ride’s statistics.
```
I'll assume you meant statistics.

**Comment by tupaschoal** [comment]
@Gymnasiast that should be it now. I have tested on a few rides and output looks ok. Used this poor plugin:

```javascript
function onParkLoad() {
    for (var ride_id = 0; ride_id < map.numRides; ride_id++) {
        var ride = map.getRide(ride_id);
        if (ride.name == "Texas Giant") {
            console.log("maxSpeed: " + ride.maxSpeed);
            console.log("averageSpeed: " + ride.averageSpeed);
            console.log("rideTime: " + ride.rideTime);
            console.log("rideLength: " + ride.rideLength);
            console.log("maxPositiveVerticalGs: " + ride.maxPositiveVerticalGs);
            console.log("maxNegativeVerticalGs: " + ride.maxNegativeVerticalGs);
            console.log("maxLateralGs: " + ride.maxLateralGs);
            console.log("totalAirTime: " + ride.totalAirTime);
            console.log("drops: " + ride.drops);
            console.log("poweredLifts: " + ride.poweredLifts);
            console.log("highestDropHeight: " + ride.highestDropHeight);
        }
    }
}

function main() {
    onParkLoad();
    if (typeof ui === 'undefined') {
        return;
    }
}

registerPlugin({
    name: 'ride_stats',
    version: '1.0',
    authors: ['Tulio Leao'],
    type: 'local',
    licence: 'MIT',
    targetApiVersion: 100,
    minApiVersion: 100,
    main: main
});
```

File: src/openrct2/scripting/bindings/ride/ScRide.cpp
Comment (Gymnasiast) [inline src/openrct2/scripting/bindings/ride/ScRide.cpp:642]: Sorry for this late addition, but:
1. Should we add `num` before these two, to make clear they are counts?
2. Digging through the code, it looks like the second property ultimately represent the number of lifts hill, whether LIM Launched or conventional chain lift. So I would propose just calling this `liftHills` or `numLiftHills`, depending on what you picked for point 1.

Comment (tupaschoal) [reply src/openrct2/scripting/bindings/ride/ScRide.cpp]: Done

**Review by Gymnasiast — APPROVED** [review]

**Pull request merged**
